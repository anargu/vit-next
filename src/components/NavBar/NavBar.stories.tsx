import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import { NavBar } from "./NavBar";

// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  title: 'UI/NavBar',
  component: NavBar,
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {
  },
} as ComponentMeta<typeof NavBar>;

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const Template: ComponentStory<typeof NavBar> = (args : any) => (
  <div className="bg-gray-100"><NavBar {...args} /></div>
);

export const Default = Template.bind({});
// More on args: https://storybook.js.org/docs/react/writing-stories/args
Default.args = {
};


