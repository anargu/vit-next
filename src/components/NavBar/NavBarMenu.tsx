import React, { useRef, useState, useEffect } from "react";
import Link from "next/link";

interface ItemSubMenu {
  label: string,
  url?: string,
  onClick?: (...args : any) => void,
}

interface NavBarItemMenuProps {
  right?: boolean,
  children: React.ReactNode,
  menu?: ItemSubMenu[],
}

export const NavBarItemMenu : React.FC<NavBarItemMenuProps> = ({ right, children, menu }) => {

  const ref = useRef<HTMLDivElement>(null);
  const [isOpen, setIsOpen] = useState(false);

  useEffect(() => {
    const handleClickOutside = (event : MouseEvent) => {
      if (ref.current && !ref.current.contains(event.target as Node)) {
        setIsOpen(false);
      }
    };

    document.addEventListener('click', handleClickOutside, true);

    return () => {
      document.removeEventListener('click', handleClickOutside, true);
    };
  }, []);

  return (
    <div className=" relative text-right" ref={ref}>
      <div onClick={() => { setIsOpen((oldValue) => !oldValue) }}>
        {children}
      </div>
      <div className={`${!isOpen && "hidden"} ${right && "right-0"}
        absolute z-30 min-w-[96px] text-base list-none bg-white rounded divide-y
        divide-gray-100 shadow dark:bg-gray-700 dark:divide-gray-600`}
      >
        {menu && (
          <ul className="py-1" aria-labelledby="dropdown">
            {menu.map((item, i) => (
              <li
                key={`nav-menu-item-${i}`}
                onClick={(e : any) => {
                  e.preventDefault?.();

                  setIsOpen(false);
                }}
              >
                {item.url
                ? <Link href={item.url}>
                  <span className="block py-2 px-4 text-sm text-gray-700 hover:bg-gray-100 dark:hover:bg-gray-600 dark:text-gray-200 dark:hover:text-white">
                    {item.label}
                  </span>
                </Link>
                : <span
                    className="block py-2 px-4 text-sm text-gray-700 hover:bg-gray-100 dark:hover:bg-gray-600 dark:text-gray-200 dark:hover:text-white"
                    onClick={(e) => {
                      e.preventDefault();

                      item.onClick?.();
                    }}
                  >
                    {item.label}
                  </span>}
                
              </li>
            ))}
          </ul>
        )}
      </div>
    </div>
  );
};

