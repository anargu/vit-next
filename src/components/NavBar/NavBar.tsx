import React, { useState } from "react";
import Image from "next/image";
import { NavBarItemMenu } from "./NavBarMenu";
import { useRouter } from "next/router";
import { AnimatePresence, motion } from "framer-motion";

export const FEED_LABEL = "Feed";
export const SAVED_LABEL = "Saved";

const variants = {
  out: {
    backgroundColor: "#FFF",
    opacity: 0,
    transition: {
      duration: 0.75
    }
  },
  in: {
    backgroundColor: "#FFF",
    transition: {
      duration: 0.75,
    }
  },
  aboutIn: {
    backgroundColor: "#d1fae5",
    transition: {
      duration: 0.75,
      delay: 1.5
    }
  }
};

interface NavBarItemProps {
  active?: boolean,
  children?: React.ReactNode,
  onClick?: () => void,
}

const NavBarItem = ({ active, children, ...props  } : NavBarItemProps) => (
  <div className={`${active ? "text-base text-black-400 underline" : "text-base text-gray-400"} cursor-pointer text-center hover:text-gray-800`} {...props}>{children}</div>
);

interface LeftItem {
  label: string,
  active: boolean,
  path: string,
}

export const NavBar = () => {
  const router = useRouter();
  const { pathname } = router;

  const [leftItems, setLeftItems] = useState<LeftItem[]>([
    { label: FEED_LABEL, path: "/", active: true, },
    { label: SAVED_LABEL, path: "/saved", active: false, },
  ]);

  const onSelectItem = (item : LeftItem) => {
    if (item.label === SAVED_LABEL) return router.push("/saved");
    if (item.label === FEED_LABEL) return router.push("/");
  };

  return (
    <AnimatePresence
      initial={false}
      mode="wait"
    >
      <motion.div
        variants={variants}
        animate={router.pathname === "/about" ? "aboutIn" : "in"}
        initial="out"
        exit="out"
        className="bg-emerald-100 grid grid-cols-[20px_48px_48px_auto_70px] lg:grid-cols-[24px_96px_96px_auto_70px] items-center h-[72px] px-4"
      >
        <Image src="/assets/v_icon_512px.png" width="24px" height="24px" sizes="(min-width: 1024px) 24px, 20px"/>
        {leftItems.map((item) =>
          <NavBarItem
            key={item.label}
            active={item.path === pathname}
            onClick={() => onSelectItem(item)}
          >{item.label}</NavBarItem>)}

        <div className="inline-block w-full"/>

        <div className="inline-grid grid-cols-1 gap-6">
          {/*
            TODO: Switch layout feature
            <Image src="/assets/mosaic.svg" width="28px" height="28px" />
          */}
          <NavBarItemMenu right menu={[
            { label: "About us", url: "/about" }
          ]}>
            <Image src="/assets/more-horizontal.svg" width="28px" height="28px" />
          </NavBarItemMenu>
        </div>
      </motion.div>
    </AnimatePresence>
  );
};

