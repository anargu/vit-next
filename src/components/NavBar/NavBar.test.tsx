import { act, fireEvent, render } from "@testing-library/react";
import { useState } from "react";
import {NavBar, FEED_LABEL, SAVED_LABEL} from "./NavBar";

jest.mock("next/router", () => ({
  __esModule: true,
  useRouter: () => {
    const [pathname, setPathname] = useState("/");

    return {
      push: (newPath : string) => { setPathname(newPath) },
      pathname,
    };
  },
}));

describe("NavBar", () => {

  it("should have two items on the left side", () => {
    const { container, getByText } = render(<NavBar/>);
    
    const feedLabel = getByText(FEED_LABEL);
    const savedLabel = getByText(SAVED_LABEL);

    expect(feedLabel).not.toBeFalsy();
    expect(savedLabel).not.toBeFalsy();

    expect(container).toMatchSnapshot();
  });

  it("has Feed Item active by default", () => {
    const { getByText } = render(<NavBar/>);

    const feedLabel = getByText(FEED_LABEL);

    expect(feedLabel.className).toContain("text-black-400");
  });

  it("set Saved Tab active when clicked", async () => {
    const { getByText } = render(<NavBar/>);

    const savedLabel = getByText(SAVED_LABEL);

    act(() => { fireEvent.click(savedLabel); });

    expect(savedLabel.className).toContain("text-black");
  });

  it("has icon menus on the right side", async () => {
    const { container } = render(<NavBar/>);

    // look for next/image component
    const items = container.firstElementChild!.querySelectorAll("span > span > img");

    expect(items.length).toBeGreaterThan(0);
  });
});

