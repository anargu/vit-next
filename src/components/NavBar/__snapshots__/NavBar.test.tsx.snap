// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar should have two items on the left side 1`] = `
<div>
  <div
    class="mocked-framer-motion-AnimatePresence"
    mode="wait"
  >
    <div
      animate="in"
      class="mocked-framer-motion-div"
      exit="out"
      initial="out"
      variants="[object Object]"
    >
      <span
        style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
      >
        <span
          style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 100% 0px 0px 0px;"
        />
        <img
          data-nimg="responsive"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
        <noscript />
      </span>
      <div
        class="text-base text-black-400 underline cursor-pointer text-center hover:text-gray-800"
      >
        Feed
      </div>
      <div
        class="text-base text-gray-400 cursor-pointer text-center hover:text-gray-800"
      >
        Saved
      </div>
      <div
        class="inline-block w-full"
      />
      <div
        class="inline-flex grid-cols-1 gap-4"
      >
        <div
          class=" relative text-right"
        >
          <div>
            <span
              class="cursor-pointer inline-flex items-center gap-2"
            >
              <svg />
            </span>
          </div>
          <div
            class="hidden right-0
        absolute z-30 min-w-[96px] text-base list-none bg-white rounded divide-y
        divide-gray-100 shadow dark:bg-gray-700 dark:divide-gray-600"
          >
            <ul
              aria-labelledby="dropdown"
              class="py-1"
            >
              <li>
                <span
                  class="block py-2 px-4 text-sm text-gray-700 hover:bg-gray-100 dark:hover:bg-gray-600 dark:text-gray-200 dark:hover:text-white"
                >
                  Sign In
                </span>
              </li>
            </ul>
          </div>
        </div>
        <div
          class=" relative text-right"
        >
          <div>
            <span
              style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
            >
              <span
                style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
              >
                <img
                  alt=""
                  aria-hidden="true"
                  src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2728%27%20height=%2728%27/%3e"
                  style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                />
              </span>
              <img
                data-nimg="intrinsic"
                decoding="async"
                src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
              />
              <noscript />
            </span>
          </div>
          <div
            class="hidden right-0
        absolute z-30 min-w-[96px] text-base list-none bg-white rounded divide-y
        divide-gray-100 shadow dark:bg-gray-700 dark:divide-gray-600"
          >
            <ul
              aria-labelledby="dropdown"
              class="py-1"
            >
              <li>
                <span
                  class="block py-2 px-4 text-sm text-gray-700 hover:bg-gray-100 dark:hover:bg-gray-600 dark:text-gray-200 dark:hover:text-white"
                >
                  About us
                </span>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
